### ISSUE TITLE
Equalizer does not appear to work

### ISSUE THREAD

**Issue by jo99jo** [segler-alex/RadioDroid#445]
Running Android 7.1.2 on Nexus6P.  Equalizer does not work in RadioDroid 2.  I just checked and  the equalizer (which looks exactly the same) works fine in Google Play Music.

**Comment by werman** [comment]
>  Equalizer does not work

Google Play Music uses external equalizer and that it works means that you have equalizer. So I assume equalizer is being successfully launched via relevant setting in RadioDroid but its effects do not apply to sound played by RadioDroid, is this correct?

In such case there is nothing that prevents it from working, at least on RadioDroid side. I hate to say this but it should work if it works for Google Play Music. 

The only thing you can do is to post logcat logs. If you don't know how but you can see e.g. https://android.gadgethacks.com/how-to/android-basics-capture-logcat-for-detailed-bug-reports-0167774/

**Comment by jo99jo** [comment]
Yes, the equalizer launches in RD2 but doesn't do anything.  There is no change to the sound whether it is off or on.

I also have another radio app on my system called AUDIALS.  It has an equalizer also but does not look like the Google Play or RD2 version.  This EQ works fine also.

The process described to get a logfile sounds way to complicated for me to accomplish.  

Am I the only person running Android with this problem?  Can you take a closer look at the code?

**Comment by werman** [comment]
> I also have another radio app on my system called AUDIALS. It has an equalizer also but does not look like the Google Play or RD2 version. This EQ works fine also.

Seems that this app has built-in equalizer. 

> Am I the only person running Android with this problem? 

There was #310.

> Can you take a closer look at the code?

The thing is that without a way for me to reproduce the problem there is not much I can do. There is a little code for this on RadioDroid side, it conforms to Android documentation and works for several people and on emulators. I'll try to search if there is something missed from the docs but don't expect anything.

Maybe you could identify what equalizer is used in your case? Maybe at least screenshot it?

**Comment by jo99jo** [comment]
I don't see any way to ID anything about the EQ.  It looks exactly the same as what is in Google Play Music, which again, works fine, so I assume it comes from base Android.  It's fairly basic as you can see.

![Screenshot_20190327-105315](https://user-images.githubusercontent.com/48929420/55101440-5ba42c80-5081-11e9-8e52-afd0a3f1c1f7.png)

**Comment by jo99jo** [comment]
Solved this problem.  Downloaded an app called Precise Volume, which contains a fully functional EQ that works with RD2 just fine.

**Comment by puppykickr** [comment]
I also have had this problem.
The Radio Droid EQ is not functional in any version or on any of my devices.

I have had Radio Droid on a 5.1 a 7.1.1 and a 8.1.0 with the EQ non-fuctional on each.

There is an EQ called Equallizer FX that has all the same functions, no ads, and works brilliantly.

I got it off of UpToDown.

**Comment by hnrkinfo** [comment]
When activating the EQ while playing a stream, I have to pause and the press play again - et violá: EQ is working

Running Android 9nnn on a Pocophone F1 (MIUI)

**Comment by werman** [comment]
Why equalizers are so wonky...

Now on my new phone I can reproduce the initial issue. It looks like equalizers may not follow the spec (see https://developer.android.com/reference/android/media/audiofx/AudioEffect.html#ACTION_DISPLAY_AUDIO_EFFECT_CONTROL_PANEL)

> The EXTRA_AUDIO_SESSION extra will indicate an audio session to which the audio effects should be applied. If no audio session is specified, either one of the follownig will happen:
> 
> - If an audio session was previously opened by the calling application with ACTION_OPEN_AUDIO_EFFECT_CONTROL_SESSION intent, the effect changes will be applied to that session. 
> - If no audio session is opened, the changes will be stored in the package specific storage area and applied whenever a new audio session is opened by this application. 

Even though we opened the session equalizer didn't recognize it and if I'm correctly reading the spec the session should be recognized by equalizer.

Adding EXTRA_AUDIO_SESSION helps in my case. However not on the master branch but on my own... I'll take a look again soon.

> When activating the EQ while playing a stream, I have to pause and the press play again - et violá: EQ is working

Uh, so much for consistency. Maybe my above finding will fix this.

**Comment by werman** [comment]
After the second look I left even more confused and my above comment is probably incorrect. It looks like working randomly... Could be tested in emulator by installing AudioFX equalizer. And I don't know who is wrong, is it RadioDroid or equalizer.

**Comment by puppykickr** [comment]
I have never been able to get the built in EQ on RadioDroid to function.

It is a common problem with many music players and apps that have them included.

Personally, I found a very versatile EQ on UpToDown called Equalizer FX.

On the site it is called 'EQ Ecualizador FX'.
Don't let the name scare you, the app is in English on my device.

It is ad free, and you can make many profiles.

It functions well with RadioDroid and all my other music apps.
